### PR TITLE
Add GPURequestAdapterOptions.allowSoftware

### PR DESCRIPTION
**Pull request by kainino0x**
Proposal for the direction I think we agreed upon; alternative to #1302, related to #1439

***

### :boom: Error: 500 Internal Server Error :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Apr 14, 2021, 6:52 PM UTC)_.

<details>
<summary>More</summary>

PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [HTML Diff Service](http://services.w3.org/htmldiff) - The HTML Diff Service is used to create HTML diffs of the spec changes suggested in a pull request.

:link: [Related URL](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fpr-preview.s3.amazonaws.com%2Fgpuweb%2Fgpuweb%2Fpull%2F1634%2F8518e80.html&doc2=https%3A%2F%2Fpr-preview.s3.amazonaws.com%2Fkainino0x%2Fgpuweb%2Fpull%2F1634.html)

```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>500 Internal Server Error</title>
</head><body>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error or
misconfiguration and was unable to complete
your request.</p>
<p>Please contact the server administrator at 
 sysreq@w3.org to inform them of the time this error occurred,
 and the actions you performed just before this error.</p>
<p>More information about this error may be available
in the server error log.</p>
</body></html>

```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20gpuweb/gpuweb%231634.)._
</details>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -1323,11 +1323,15 @@ configuration is suitable for the application.
 
 <script type=idl>
 dictionary GPURequestAdapterOptions {
+    GPUAllowSoftwareOption allowSoftware = "last-resort";
     GPUPowerPreference powerPreference;
 };
-</script>
 
-<script type=idl>
+enum GPUAllowSoftwareOption {
```
Comment (kvark) [inline spec/index.bs:1330]: Why can't this just be a boolean? It's up to the UA to figure out when this should be preferred, anyway.

File: spec/index.bs [hunk 2]
```diff
@@ -1337,6 +1341,33 @@ enum GPUPowerPreference {
...
+        This behavior is implementation-defined: user agents may not always obey this option.
+        For example, if a system has very limited GPU hardware but a fast software implementation,
+        the user agent may always prefer the software implementation.
+
+        It must be one of the following <dfn enum for=>GPUAllowSoftwareOption</dfn> values:
+
+        <dl dfn-type=enum-value dfn-for=GPUAllowSoftwareOption>
+            : <dfn>"never"</dfn>
+            ::
+                Adapters with major performance caveats should never be allowed.
```
Comment (kvark) [inline spec/index.bs:1358]: still talks about the major performance caveats...

**Review by kvark — CHANGES_REQUESTED** [review]

File: spec/index.bs (hunk 1 above)
Comment (kainino0x) [reply spec/index.bs:1330]: For extensibility, see [1]

File: spec/index.bs [hunk 3]
```diff
@@ -1337,6 +1341,33 @@ enum GPUPowerPreference {
...
+            : <dfn>"last-resort"</dfn>
+            ::
+                Adapters with major performance caveats should only be allowed
+                if no other adapters are available.
+        </dl>
+
+        Issue:
+        Additional possible options here include `"always"`, if requestAdapter is pluralized,
+        and `"only"`, if we want to allow apps to force software
+        (e.g. as a driver bug workaround, or for automated testing).
```
Comment (kainino0x) [inline spec/index.bs:1369]: [1]

File: spec/index.bs (hunk 2 above)
Comment (kainino0x) [reply spec/index.bs:1358]: oops.

File: spec/index.bs (hunk 1 above)
Comment (kvark) [reply spec/index.bs:1330]: I think, if you want to have rich controls like this, then the "allow" part becomes confusing. I.e. `allowSoftware = "always"` doesn't read to me like a thing that would *prefer* software over hardware adapters.

Comment (kainino0x) [reply spec/index.bs:1330]: with plural requestAdapters, I envisioned "always" would still list software adapters last, effectively keeping the same preference order but not hiding software adapters.

**Comment by kainino0x** [comment]
Editors meeting resolution: This solution is not ideal as it doesn't actually solve all cases - for example, it doesn't allow an application to implement logic to blocklist known buggy GPUs and fall back to software (or to another GPU in the system).
It is also rather hamfisted with the 3-4 possible "allowSoftware" modes.

Plural `requestAdapters()`, plus a `GPUAdapter.isSoftware` flag, would solve this.
We can expose both hardware and software adapters to a page and allow them to choose.
Additionally, it would resolve @jdashg's concern that we make it too easy for applications to disallow software implementations 'just because it seems nice'.

@toji is looking at a proposal that incorporates plural `requestAdapters()` + `.isSoftware` together.
Exposing both hardware and software adapters to a page is 
This doesn't have to expose multiple hardware adapters (and in fact we could require, say, a Permissions API permission in order to see multiple hardware adapters).

**Comment by kainino0x** [comment]
Closing in favor of #1734

**Pull request closed without merging**